[FLINK-33274][release] Add release note for version 1.18 (#23527)
apache/flink: docs/content.zh/release-notes/flink-1.18.md

regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

# Release notes - Flink 1.18

These release notes discuss important aspects, such as configuration, behavior or dependencies,
that changed between Flink 1.17 and Flink 1.18. Please read these notes carefully if you are
planning to upgrade your Flink version to 1.18.

### Build System

#### Support Java 17 (LTS)

##### [FLINK-15736](https://issues.apache.org/jira/browse/FLINK-15736)
Apache Flink was made ready to compile and run with Java 17 (LTS). This feature is still in beta mode. 
Issues should be reported in Flink's bug tracker.

### Table API & SQL

> #### Unified the max display column width for SqlClient and Table APi in both Streaming and Batch execMode

Reviewer: ```suggestion
#### Unified the max display column width for SQL Client and Table API in both Streaming and Batch execution mode
```

##### [FLINK-30025](https://issues.apache.org/jira/browse/FLINK-30025)

> Introduction of the new ConfigOption DISPLAY_MAX_COLUMN_WIDTH (table.display.max-column-width) 

Reviewer 2: ```suggestion
Introduction of the new ConfigOption `DISPLAY_MAX_COLUMN_WIDTH` (`table.display.max-column-width`) 
```
just add code marker

in the TableConfigOptions class is now in place. 

> This option is utilized when displaying table results through the Table API and sqlClient. 

Reviewer 2: ```suggestion
This option is utilized when displaying table results through the Table API and SQL Client. 
```
I noticed in docs that most of the time it is either `SQL Client` or `sql client`, probably would make sense to follow same principle

> As sqlClient relies on the Table API underneath, and both sqlClient and the Table API serve distinct 

Reviewer 2: ```suggestion
As sqlClient relies on the Table API underneath, and both SQL Client and the Table API serve distinct 
```

and isolated scenarios, it is a rational choice to maintain a centralized configuration. 

> This approach also simplifies matters for users, as they only need to manage one configOption for display control.

Reviewer 3: ```suggestion
This approach also simplifies matters for users, as they only need to manage one ConfigOption for display control.
```

> During the migration phase, while sql-client.display.max-column-width is deprecated, 

Reviewer 2: ```suggestion
During the migration phase, while `sql-client.display.max-column-width` is deprecated, 
```
just add code marker

> any changes made to sql-client.display.max-column-width will be automatically transferred to table.display.max-column-width. 

Reviewer 2: ```suggestion
any changes made to `sql-client.display.max-column-width` will be automatically transferred to `table.display.max-column-width`. 
```
just add code marker

Caution is advised when using the CLI, as it is not recommended to switch back and forth between these two options.

#### Introduce Flink Jdbc Driver For Sql Gateway
##### [FLINK-31496](https://issues.apache.org/jira/browse/FLINK-31496)

> Apache Flink now supports JDBC driver to access sql-gateway, you can use the driver in any cases that

Reviewer: ```suggestion
Apache Flink now supports JDBC driver to access SQL Gateway, you can use the driver in any cases that
```

support standard JDBC extension to connect to Flink cluster.

#### Extend watermark-related features for SQL
##### [FLINK-31535](https://issues.apache.org/jira/browse/FLINK-31535)
Flink now enables user config watermark emit strategy/watermark alignment/watermark idle-timeout

> in Flink sql job with dynamic table options and 'Options' hint.

Reviewer 2: ```suggestion
in Flink sql job with dynamic table options and `OPTIONS` hint.
```
usually hints are referenced in doc in uppercase + added code marker

#### Support configuring CatalogStore in Table API
##### [FLINK-32431](https://issues.apache.org/jira/browse/FLINK-32431)
Support lazy initialization of catalog and persistence of catalog configuration.

#### Deprecate ManagedTable related APIs
##### [FLINK-32656](https://issues.apache.org/jira/browse/FLINK-32656)
ManagedTable related APIs are deprecated and will be removed in a future major release.

### Connectors & Libraries

> #### SplitReader doesn't extend AutoCloseable but implements close

Reviewer 3: ```suggestion
#### SplitReader implements AutoCloseable instead of providing its own close method
```
I see here is the issue title. But I think we should describe it as this in the release note.

Previously such large jobs were actually at large risk of overloading JobManager, now that's far less likely to happen.

#### Replace Akka by Pekko
##### [FLINK-32468](https://issues.apache.org/jira/browse/32468)
Flink's RPC framework is now based on Apache Pekko instead of Akka. Any Akka dependencies were removed.

#### Introduce Runtime Filter for Flink Batch Jobs
##### [FLINK-32486](https://issues.apache.org/jira/browse/FLINK-32486)
We introduced a runtime filter for batch jobs in 1.18, which is designed to improve join performance. 
It will dynamically generate filter conditions for certain Join queries at runtime to reduce the amount of scanned or shuffled data, 
avoid unnecessary I/O and network transmission, and speed up the query. 
Its working principle is building a filter(e.g. bloom filter) based on the data on the small table side(build side) first, 
then passing this filter to the large table side(probe side) to filter the irrelevant data on it, 
this can reduce the data reaching the join and improve performance.

#### Make watermark alignment ready for production use
##### [FLINK-32548](https://issues.apache.org/jira/browse/FLINK-32548)

> The watermark alignment is ready for production since Flink-1.18, 

Reviewer 3: ```suggestion
The watermark alignment is ready for production since Flink 1.18, 
```

which completed a series of bug fixes and improvements related to watermark alignment. 
Please refer to [FLINK-32420](https://issues.apache.org/jira/browse/FLINK-32420) for more information.

> #### Redundant taskManagers should always be fulfilled in FineGrainedSlotManager

Reviewer 4: ```suggestion
#### Redundant TaskManagers should always be fulfilled in FineGrainedSlotManager
```

##### [FLINK-32880](https://issues.apache.org/jira/browse/FLINK-32880)

> Fix the issue that redundant taskManagers will not be fulfilled in FineGrainedSlotManager periodically.

Reviewer 4: ```suggestion
Fix the issue that redundant TaskManagers will not be fulfilled in FineGrainedSlotManager periodically.
```

#### RestClient can deadlock if request made after Netty event executor terminated
##### [FLINK-32583](https://issues.apache.org/jira/browse/FLINK-32583)
Fix a bug in the RestClient where making a request after the client was closed returns a future that never completes.

#### Deprecate Queryable State
##### [FLINK-32559](https://issues.apache.org/jira/browse/FLINK-32559)
The Queryable State feature is formally deprecated. It will be removed in future major version bumps.

### SDK

#### Properly deprecate DataSet API
##### [FLINK-32558](https://issues.apache.org/jira/browse/FLINK-32558)
DataSet API is formally deprecated, and will be removed in the next major release.

### Dependency upgrades

#### Upgrade Calcite version to 1.32.0
##### [FLINK-27998]
##### [FLINK-28744]
##### [FLINK-29319]
Due to CALCITE-4861 (Optimization of chained CAST calls can lead to unexpected behavior), 
also Flink's casting behavior has slightly changed. Some corner cases might behave differently now: For example, 

> casting from FLOAT/DOUBLE 9234567891.12 to INT/BIGINT has now Java behavior for overflows.

Reviewer 2: ```suggestion
casting from `FLOAT`/`DOUBLE` 9234567891.12 to `INT`/`BIGINT` has now Java behavior for overflows.
```
just adde code mark
